linkerd/stack/src/fail_on_error.rs

```diff
@@ -0,0 +1,88 @@
+use linkerd2_error::Error;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::Mutex;
+
+#[derive(Clone, Debug)]
+pub struct FailOnError<E, S> {
+    inner: S,
+    error: Arc<Mutex<Option<SharedError>>>,
+    _marker: std::marker::PhantomData<E>,
+}
+
+#[derive(Clone, Debug)]
+struct SharedError(Arc<Error>);
+
+impl<E, S> FailOnError<E, S> {
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner,
+            error: Arc::new(Mutex::new(None)),
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<E, S, Req> tower::Service<Req> for FailOnError<E, S>
+where
+    S: tower::Service<Req>,
+    S::Response: Send,
+    S::Error: Into<Error> + Send,
+    S::Future: Send + 'static,
+    E: std::error::Error + 'static,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        futures::ready!(self.inner.poll_ready(cx).map_err(Into::into))?;
+        if let Ok(e) = self.error.try_lock() {
+            if let Some(e) = &*e {
+                return Poll::Ready(Err(e.clone().into()));
+            }
```

Comment:
It seems like we may want to use a `RwLock`, so that this case is not synchronized?  It seems like it's possible for `poll_ready` to spuriously return `Ok` if another thread is currently accessing the lock. If we used a `RwLock` here we would be able to clone the error even if someone else also is...